week9/answers.sql

```diff
@@ -1,3 +1,195 @@
+ );
+ 
+  CREATE TABLE friends (
+    user_friend_id INT UNSIGNED PRIMARY KEY NOT NULL AUTO_INCREMENT,
+    user_id INT UNSIGNED NOT NULL,
+    friend_id INT UNSIGNED NOT NULL,
+		FOREIGN KEY (user_id)
+        REFERENCES users (user_id)
+        ON UPDATE CASCADE
+        ON DELETE CASCADE,
+		FOREIGN KEY (friend_id)
+		REFERENCES users (user_id)
+        ON UPDATE CASCADE
+        ON DELETE CASCADE
+ );
+ 
+   CREATE TABLE posts (
+    post_id INT UNSIGNED PRIMARY KEY NOT NULL AUTO_INCREMENT,
+    user_id INT UNSIGNED NOT NULL,
+    created_on TIMESTAMP NOT NULL DEFAULT NOW(),
+    updated_on TIMESTAMP NOT NULL DEFAULT NOW() ON UPDATE NOW(),
+    content VARCHAR(250) NOT NULL,
+		FOREIGN KEY (user_id)
+		REFERENCES users (user_id)
+		ON UPDATE CASCADE
+		ON DELETE CASCADE
+ );
+ 
+    CREATE TABLE notifications (
+    notification_id INT UNSIGNED PRIMARY KEY NOT NULL AUTO_INCREMENT,
+    post_id INT UNSIGNED NOT NULL,
+    user_id INT UNSIGNED NOT NULL,
+		FOREIGN KEY (user_id)
+		REFERENCES users (user_id)
+		ON UPDATE CASCADE
+		ON DELETE CASCADE,
+        FOREIGN KEY (post_id)
+		REFERENCES posts (post_id)
+		ON UPDATE CASCADE
+		ON DELETE CASCADE
+ );
+ 
+ 
+ 
+-- ------------------ Notification Posts View ------------------ 
+
+CREATE OR REPLACE VIEW notification_posts AS 
+	SELECT 
+			n.user_id AS user_id,
+			u.first_name AS first_name,
+			u.last_name AS last_name,
+            p.post_id AS post_id,
+		 	p.content AS content
+		FROM notifications n
+			INNER JOIN posts p
+				ON p.post_id = n.post_id
+			RIGHT OUTER JOIN users u
+				ON u.user_id = p.user_id;
+        
+        
+        
+-- ------------------ Notify All Procedure ------------------ 
+
+DELIMITER ;;
+CREATE PROCEDURE notify_all(this_post_id INT UNSIGNED)
+
+BEGIN
+
+	-- Get count of all users
+    DECLARE user_count INT;
+    DECLARE cur_user INT;
+    
+	SELECT COUNT(user_id) FROM users INTO user_count;
+    SET cur_user = 1;
+	
+    -- Loop through all users and add notification for them
+    WHILE cur_user < user_count DO
+		INSERT INTO notifications (user_id, post_id) VALUES (cur_user, this_post_id);
+        SET cur_user = cur_user + 1;
+	END WHILE;
+     
+END;;
+DELIMITER ;
+
+
+
+ 
+-- ------------------ Add Post Procedure ------------------
+
+DELIMITER ;;
+CREATE PROCEDURE add_post(this_user_id INT UNSIGNED, this_content VARCHAR(250))
+BEGIN
+    DECLARE this_post_id INT;
+    DECLARE cur_friend INT;
+    DECLARE row_not_found TINYINT DEFAULT FALSE;
+    
+    --  Friends Cursor
+    DECLARE friends_cursor CURSOR FOR 
+		SELECT f.friend_id
+			FROM friends f
+            WHERE f.user_id = this_user_id
+			ORDER BY f.friend_id ASC;
+            
+	-- Friends Cursor Row Not Found Handler
+    DECLARE CONTINUE HANDLER FOR NOT FOUND
+		SET row_not_found = TRUE;
+    
+    -- Make Post
+    INSERT INTO posts (user_id, content) VALUES (this_user_id, this_content);
+
+    -- Get current post ID
+    SELECT LAST_INSERT_ID() FROM posts LIMIT 1 INTO this_post_id;
```

Comment:
It's giving me this error when I don't have LIMIT 1:

Error Code: 1172. Result consisted of more than one row 0.000 sec

Reply:
Try simply:

```sql
SELECT LAST_INSERT_ID() INTO this_post_id;
```

LAST_INSERT_ID() operates on a connection, not based on a selected table (so, if you happened to do your last insert in the `users` table and ran this query, it would still only give you the ID from the `users` table despite the `SELECT ... FROM posts`.